pkg/controller/install/rule_checker.go

```diff
@@ -0,0 +1,140 @@
+package install
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	rbacauthorizer "k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient"
+)
+
+// RuleChecker is used to verify whether PolicyRules are satisfied by existing Roles or ClusterRoles
+type RuleChecker interface {
+	// RuleSatisfied determines whether a PolicyRule is satisfied for a ServiceAccount
+	// by existing Roles and ClusterRoles
+	RuleSatisfied(sa *corev1.ServiceAccount, namespace string, rule rbacv1.PolicyRule) (bool, error)
+}
+
+// CSVRuleChecker determines whether a PolicyRule is satisfied for a ServiceAccount
+// by existing Roles and ClusterRoles
+type CSVRuleChecker struct {
+	opClient operatorclient.ClientInterface
+	csv      *v1alpha1.ClusterServiceVersion
+}
+
+func NewCSVRuleChecker(opClient operatorclient.ClientInterface, csv *v1alpha1.ClusterServiceVersion) *CSVRuleChecker {
+	return &CSVRuleChecker{
+		opClient: opClient,
+		csv:      csv.DeepCopy(),
+	}
+}
+
+// RuleSatisfied returns true if a ServiceAccount is authorized to perform all actions described by a PolicyRule in a namespace
+func (c *CSVRuleChecker) RuleSatisfied(sa *corev1.ServiceAccount, namespace string, rule rbacv1.PolicyRule) (bool, error) {
+	// get attributes set for the given Role and ServiceAccount
+	user := toDefaultInfo(sa)
+	attributesSet := toAttributesSet(user, namespace, rule)
+
+	// create a new RBACAuthorizer
+	rbacAuthorizer := rbacauthorizer.New(c, c, c, c)
```

Comment:
I think this is pretty important: https://github.com/kubernetes/kubernetes/blob/473ebb21d168b74c1d6d729faf513ad311d30e2b/pkg/kubeapiserver/authorizer/config.go#L108-L113

I think we want to wire up to informers like this, so that we pull from caches instead of querying the api many times

Reply:
Agreed. I'm thinking that I'll need to write wrappers around each of the listers to filter on ownerref too.